### PR TITLE
Json processor: add_to_root_recursive_merge option

### PR DESCRIPTION
**Pull request by felixbarny**
This is to support creating an ingest pipeline for ECS JSON logs.

Merging instead of `putAll` is important as Filebeat sets values for `data_stream.type`, `data_stream.dataset`, and `data_stream.namespace`. If the logs override just one, such as `data_stream.dataset`, currently, the whole `data_stream` namespace would be overridden.

### CODE REVIEW

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/JsonProcessor.java [hunk 1]
```diff
@@ -117,17 +140,23 @@ public JsonProcessor create(Map<String, Processor.Factory> registry, String proc
...
             boolean addToRoot = ConfigurationUtils.readBooleanProperty(TYPE, processorTag, config, "add_to_root", false);
+            boolean addToRootRecursiveMerge = ConfigurationUtils.readBooleanProperty(TYPE, processorTag, config,
+                "add_to_root_recursive_merge", false);
 
             if (addToRoot && targetField != null) {
                 throw newConfigurationException(TYPE, processorTag, "target_field",
                     "Cannot set a target field while also setting `add_to_root` to true");
             }
+            if (addToRoot == false && addToRootRecursiveMerge) {
+                throw newConfigurationException(TYPE, processorTag, "add_to_root_recursive_merge",
```
Comment (martijnvg) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/JsonProcessor.java:151]: maybe also add a unit test to see that this fails?

File: docs/reference/ingest/processors/json.asciidoc [hunk 2]
```diff
@@ -10,10 +10,11 @@ Converts a JSON string into a structured JSON object.
...
-| `target_field` | no        | `field`  | The field that the converted structured object will be written into. Any existing content in this field will be overwritten.
-| `add_to_root`  | no        | false    | Flag that forces the serialized json to be injected into the top level of the document. `target_field` must not be set when this option is chosen.
+| Name                         | Required  | Default  | Description
+| `field`                      | yes       | -        | The field to be parsed.
+| `target_field`               | no        | `field`  | The field that the converted structured object will be written into. Any existing content in this field will be overwritten.
+| `add_to_root`                | no        | false    | Flag that forces the serialized json to be merged into the top level of the document. `target_field` must not be set when this option is chosen.
+| `add_to_root_recursive_merge`| no        | false    | Flag that controls the merging strategy for `add_to_root`. When set to `false`, only the first level will be merged. When set to `true`, even nested objects will be merged. Only applicable if `add_to_root` is set to `true`.
```
Comment (martijnvg) [inline docs/reference/ingest/processors/json.asciidoc:17]: I'm not so sure about the name. Not sure what a better name is. Maybe `add_to_root_merge_strategy` with values `recursive` and `replace`? I curious to see what others think.

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/JsonProcessor.java (hunk 1 above)
Comment (felixbarny) [reply modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/JsonProcessor.java:151]: done

File: docs/reference/ingest/processors/json.asciidoc (hunk 2 above)
Comment (felixbarny) [reply docs/reference/ingest/processors/json.asciidoc:17]: +1 I like the descriptiveness of the values.
I've created another PR that implements that: https://github.com/elastic/elasticsearch/pull/74967. Feel free to close this PR in favor of the `add_to_root_merge_strategy` PR.

**Comment by felixbarny** [comment]
Closing in favor of https://github.com/elastic/elasticsearch/pull/74967

**Pull request closed without merging**